sa/model_test.go

```diff
@@ -224,3 +239,144 @@ func TestPopulateAttemptedFieldsBadJSON(t *testing.T) {
+
+	// Register and insert a certificate with a serial of `1337`.
+	testCtx.setupCertificate1337A(t)
+
+	// This should return the certificate that we just inserted.
+	_, err := SelectCertificate(testCtx.db, serialString)
+	test.AssertNotError(t, err, "received an error for a valid query")
+
+	// Register and insert a certificate with a serial of `1337` but for a
+	// different hostname.
+	testCtx.setupCertificate1337B(t)
+
+	// With a duplicate being present, this should error.
+	_, err = SelectCertificate(testCtx.db, serialString)
+	test.AssertError(t, err, "should've received an error for multiple rows")
+}
+
+type testCtx struct {
+	db      *db.WrappedMap
+	ssa     core.StorageAdder
+	cleanUp func()
+}
+
+func (c testCtx) registerCertificate(t *testing.T, emailRaw string, jsonKey []byte, regID int64) {
+	var key jose.JSONWebKey
+	err := json.Unmarshal(jsonKey, &key)
+	test.AssertNotError(t, err, "Failed to unmarshal public JWK")
+
+	var reg core.Registration
+	email := "mailto:" + emailRaw
+	reg = core.Registration{
+		ID: regID,
+		Contact: &[]string{
+			email,
+		},
+		Key:       &key,
+		InitialIP: net.ParseIP("127.0.0.1"),
+	}
+	ctx := context.Background()
+	_, err = c.ssa.NewRegistration(ctx, reg)
+	test.AssertNotError(t, err, "Couldn't complete registration")
+}
+
+func (ctx testCtx) insertCertificate(t *testing.T, hostname string, serial, regID int64) {
+	serialBigInt := big.NewInt(serial)
+	serialString := core.SerialToString(serialBigInt)
+
+	fc := newFakeClock(t)
+
+	rawCert := x509.Certificate{
+		Subject: pkix.Name{
+			CommonName: "leet",
+		},
+		NotAfter:     fc.Now().Add(30 * 24 * time.Hour),
+		DNSNames:     []string{hostname},
+		SerialNumber: serialBigInt,
+	}
+
+	n := bigIntFromB64("n4EPtAOCc9AlkeQHPzHStgAbgs7bTZLwUBZdR8_KuKPEHLd4rHVTeT-O-XV2jRojdNhxJWTDvNd7nqQ0VEiZQHz_AJmSCpMaJMRBSFKrKb2wqVwGU_NsYOYL-QtiWN2lbzcEe6XC0dApr5ydQLrHqkHHig3RBordaZ6Aj-oBHqFEHYpPe7Tpe-OfVfHd1E6cS6M1FZcD1NNLYD5lFHpPI9bTwJlsde3uhGqC0ZCuEHg8lhzwOHrtIQbS0FVbb9k3-tVTU4fg_3L_vniUFAKwuCLqKnS2BYwdq_mzSnbLY7h_qixoR7jig3__kRhuaxwUkRz5iaiQkqgc5gHdrNP5zw==")
+	e := intFromB64("AQAB")
+	d := bigIntFromB64("bWUC9B-EFRIo8kpGfh0ZuyGPvMNKvYWNtB_ikiH9k20eT-O1q_I78eiZkpXxXQ0UTEs2LsNRS-8uJbvQ-A1irkwMSMkK1J3XTGgdrhCku9gRldY7sNA_AKZGh-Q661_42rINLRCe8W-nZ34ui_qOfkLnK9QWDDqpaIsA-bMwWWSDFu2MUBYwkHTMEzLYGqOe04noqeq1hExBTHBOBdkMXiuFhUq1BU6l-DqEiWxqg82sXt2h-LMnT3046AOYJoRioz75tSUQfGCshWTBnP5uDjd18kKhyv07lhfSJdrPdM5Plyl21hsFf4L_mHCuoFau7gdsPfHPxxjVOcOpBrQzwQ==")
+	p := bigIntFromB64("uKE2dh-cTf6ERF4k4e_jy78GfPYUIaUyoSSJuBzp3Cubk3OCqs6grT8bR_cu0Dm1MZwWmtdqDyI95HrUeq3MP15vMMON8lHTeZu2lmKvwqW7anV5UzhM1iZ7z4yMkuUwFWoBvyY898EXvRD-hdqRxHlSqAZ192zB3pVFJ0s7pFc=")
+	q := bigIntFromB64("uKE2dh-cTf6ERF4k4e_jy78GfPYUIaUyoSSJuBzp3Cubk3OCqs6grT8bR_cu0Dm1MZwWmtdqDyI95HrUeq3MP15vMMON8lHTeZu2lmKvwqW7anV5UzhM1iZ7z4yMkuUwFWoBvyY898EXvRD-hdqRxHlSqAZ192zB3pVFJ0s7pFc=")
+	testKey := rsa.PrivateKey{PublicKey: rsa.PublicKey{N: n, E: e}, D: d, Primes: []*big.Int{p, q}}
```

Comment:
Let's factor this out into a `func makeTestKey`. I'm surprised we didn't already have one in SA!